packages/core/saved-objects/core-saved-objects-migration-server-internal/src/document_migrator/pipelines/downgrade_pipeline.ts

```diff
@@ -0,0 +1,137 @@
+        throw new Error(`Changing a document's type during a migration is not supported.`);
+      }
+      this.document = applyVersion({ document: transformedDoc, transform });
+    }
+
+    this.document = this.ensureVersion(this.document);
+
+    return {
+      document: this.document,
+      additionalDocs: [],
+    };
+  }
+
+  private getPendingTransforms() {
+    const { transforms } = this.typeTransforms;
+    return transforms.reverse().filter((transform) => this.isPendingTransform(transform));
+  }
+
+  private isPendingTransform({ transformType, version }: Transform) {
+    const { coreMigrationVersion, typeMigrationVersion } = this.document;
+
+    switch (transformType) {
+      // reference and convert type transforms were deprecated before downward conversion were a thing
+      // so we will never need to revert such type of transforms (and they didn't implement a down conversion)
+      case TransformType.Reference:
+      case TransformType.Convert:
+        return false;
+      // including core transforms if targetCoreVersion is specified
+      case TransformType.Core:
+        return (
+          this.targetCoreVersion &&
+          (coreMigrationVersion == null ||
+            Semver.gt(coreMigrationVersion, this.targetCoreVersion)) &&
+          Semver.gt(version, this.targetCoreVersion)
+        );
+      // including migrate transforms between the targetTypeVersion and the typeMigrationVersion
+      case TransformType.Migrate:
+        return (
+          (typeMigrationVersion == null || Semver.gte(typeMigrationVersion, version)) &&
+          Semver.gt(version, this.targetTypeVersion)
+        );
+    }
+  }
+
+  /**
+   * Verifies that the current document version is not greater than the version supported by Kibana.
+   * And that the targetTypeVersion is not greater than the document's
+   */
+  private assertCompatibility() {
+    const { id, typeMigrationVersion: currentVersion } = this.document;
+    const latestVersion = this.typeTransforms.latestVersion.migrate;
+
+    if (currentVersion && Semver.gt(currentVersion, latestVersion)) {
+      throw Boom.badData(
+        `Document "${id}" belongs to a more recent version of Kibana [${currentVersion}] when the last known version is [${latestVersion}].`,
+        this.document
+      );
+    }
+
+    if (currentVersion && Semver.gt(this.targetTypeVersion, currentVersion)) {
+      throw new Error(
+        `Trying to transform down to a higher version: ${currentVersion} to ${this.targetTypeVersion}`
+      );
+    }
+  }
+
+  private ensureVersion({
+    coreMigrationVersion: currentCoreMigrationVersion,
+    typeMigrationVersion: currentTypeMigrationVersion,
+    ...document
+  }: SavedObjectUnsanitizedDoc) {
+    const coreMigrationVersion = this.targetCoreVersion || currentCoreMigrationVersion;
+
+    return {
+      ...document,
+      typeMigrationVersion: this.targetTypeVersion,
+      ...(coreMigrationVersion ? { coreMigrationVersion } : {}),
```

Comment:
Different behavior than the up migrations, here we always apply/set the target version to the document before returning it.